### ISSUE TITLE
Sending packets in parallel - worth it?

### ISSUE THREAD

**Issue by naddika** [robertdavidgraham/masscan#500]
I haven't examined the code yet. As I understand, it's single threaded, right?

If so, provided that by default masscan is already quite fast and can end up using up all the bandwidth of a network, even if network is wide and fast, would it make the project even faster if it was parallel? If it **send and receive packets in parallel**, that is.

If time and effort necessary to implement that weren't a factor.

**Comment by mzpqnxow** [comment]
It is single threaded but it also has an explicitly optimized tx loop and is asynchronous. If you want to increase the packet rate, the correct way is to use Linux with a zero-copy/RDMA driver that enables the NIC to more efficiently bypass the kernel, which is where much of the overhead in packet transmission exists. There are two technologies I'm aware of that allow you to do this and I've used both- they work

- [DPDK](https://www.dpdk.org/) supports many different brands and models of 1Gbps, 10Gbps and 40Gbps NICs. Some cloud platforms such as Azure even [support this](https://docs.microsoft.com/en-us/azure/virtual-network/setup-dpdk) when Accelerated Networking is enabled on the Virtual NIC. Accelerated Networking provides a Mellanox device to the VM that is supported by DPDK
- [PF_RING Zero Copy](https://www.ntop.org/products/packet-capture/pf_ring/pf_ring-zc-zero-copy/) accomplishes the same thing but it only supports Intel 1Gbps, 10Gbps and 40Gbps NICs

I don't know what other options there are, but masscan can make use of both of this without any special patches. With this the only limitation is your upstream pipe and infrastructure. With a Gigabit burstable pipe you can easily send well over 1mm pps with a single thread and a 1Gbps NIC

So in short, masscan is by design both the fastest option while also being intentionally "limited" to 1 thread. The correct way to get better performance is to utilize these frameworks. If you want to push the limits further, you should use separate processes with a dedicated interface for each process. You'll also want to specify a different `--adapter-port` for each process. You can utilize the native sharding support provided by masscan

For more details, you should see the [README](https://github.com/robertdavidgraham/masscan/blob/master/README.md), specifically the sections `Randomization`, `Asynchronous` and `C10 Scalability`

Hope this is helpful

**Comment by naddika** [comment]
Thanks, let me see.
